src/app/clusters/ota-requestor/OTARequestorStorageImpl.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "OTARequestorStorageImpl.h"
+
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <platform/OTARequestorInterface.h>
+
+#include <limits>
+
+namespace chip {
+
+CHIP_ERROR OTARequestorStorageImpl::StoreDefaultProvider(const ProviderLocationType & provider)
+{
+    uint8_t buffer[kMaxSerializedSize];
+    TLV::TLVWriter writer;
+
+    writer.Init(buffer);
+    ReturnErrorOnFailure(provider.EncodeForRead(writer, TLV::AnonymousTag(), provider.fabricIndex));
+
+    return mPersistentStorage->SyncSetKeyValue(DefaultStorageKeyAllocator().OTADefaultProvider(provider.fabricIndex), buffer,
+                                               static_cast<uint16_t>(writer.GetLengthWritten()));
+}
+
+CHIP_ERROR OTARequestorStorageImpl::ClearDefaultProvider(FabricIndex fabricIndex)
+{
+    return mPersistentStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator().OTADefaultProvider(fabricIndex));
+}
+
+CHIP_ERROR OTARequestorStorageImpl::LoadDefaultProviders(ProviderLocationList & providers)
```

Comment:
This will actually cause re-ordering of the entries in the list, to be in order of fabricIndex, as opposed to the order read. This may run afoul of list stability rules for non-fabric-scoped reads. This is also quite costly. If FabricIndex had been 16 bits, we would have right-away said "you can't exhaustively try 65536 values". I think this applies here as well.

This is solved in other clusters by a variety of means, but I would recommend that we at least have an "index" data structure written to keep track of which positional entries relate to which fabric. Just a basic list of fabric index in order, and then the entry can be looked-up by FabricIndex.

Alternatively, the entire structure could be stored since it's likely not huge (min num fabrics = 5 in spec), so the list would be a TLV array with N=5 entries of (8bit fabric index, 64 bits nodeid, 16 bits endpoint id,  + tags) so ~5 * (~16 bytes), which is overall reasonably small to write as a unit every time the list is mutated. That would also save persistent storage slots.

Reply:
> This may run afoul of list stability rules for non-fabric-scoped reads.

Reply:
The only reason I wanted to use a separate slot for a specific fabric is to reduce the number of written bytes if a single fabric is modifed, so using another list to maintain indices doesn't make much sense from that perspective. I'll encode the entire list, but we'll then need to expand the attribute accessor interface to know when all list items have been parsed because currently, they're added one-by-one.

Reply:
@tcarmelveilleux  PTAL